### ISSUE TITLE
Completing XOVERLAPPED before returning from export causes issues in some games

### ISSUE THREAD

**Issue by emoose** [xenia-project/xenia#1296]
Right now all kinds of exports that take XOVERLAPPED params are completing them before returning from the export, this works fine in most cases but unfortunately seems to cause issues in others, where I guess maybe the game isn't expecting them to be immediately completed and ends up not running correctly as a result.

Some of the worst examples are probably the XamShow*UI functions, in those we actually wait for user input before returning from the export at all, meaning we could be stalling the game for possibly minutes/hours before we return control to it. (while the stalling is maybe a separate issue, it can probably be fixed with the same solution to the overlapped problem)

An easy way I found to remedy this is by putting the main code (+ the overlapped completion code) of the export into a separate thread (or if the main code is pretty much non-time-consuming, having a small thread sleep just before completion), that way the overlapped is still in progress when the game regains execution, and the export also takes a fraction of the time it previously took to run too. (changes for this are in my branch here: https://github.com/emoose/xenia/commits/xamshowui-fixes, though in XamShowDeviceSelectorUI I use an imgui dialog to simulate a thread sleep since I didn't know how to actually make the thread sleep at the time :P)

This solved issues in the dash.xex during initial setup where it wouldn't recognise input from XamShowKeyboardUI, also fixes Sonic Unleashed not recognising the device id returned from XamShowDeviceSelectorUI, and Project Sylpheed also now goes in-game with the fix, there's probably many other games where this helps with things too.

I'm not sure if creating new threads for every export that uses XOVERLAPPED params is really the best idea though, I could see threads quickly piling up, maybe even ending up in contention with each other...
IIRC 360's Xam uses a task-queue system instead, where tasks are done sequentially rather than my current solution of starting a new thread immediately.
Something like that would probably be a lot better, and bring us closer to how the 360 actually works, though I'm not really sure where to start with something like that...

**Comment by illusion98** [comment]
This fix also gets Warriors Orochi 3 in-game
![](https://cdn.discordapp.com/attachments/308207592482668545/595179441378820097/Untitled.png)

**Comment by emoose** [comment]
Saw some talk on discord about Project Sylpheed, so bumping this issue since AFAIK this is what  causes it to freeze with a "Select storage device" window ingame.

My threaded-xoverlapped branch seems to fix this, you can see the changed code for XamShowDeviceSelectorUI here:
https://github.com/emoose/xenia/commit/e4fbe611f9c65bcf0fec992309fc615cd7dc817d#diff-d057ca15d4ce4bef657c34835cd0300eR358

All that does is spawn a new thread to handle completing the xoverlapped and then immediately returns from XamShowDeviceSelectorUI back to game code, with the XOVERLAPPED set to pending.
The thread then waits 1.5s before completing the xoverlapped (and running the overlapped's completion handler, etc)

In other words, instead of the old way:
```Game -> XamShowDeviceSelectorUI -> Games XOVERLAPPED completion handler -> Game```
It would now work like
```Game -> XamShowDeviceSelectorUI -> Game```
```Seperate thread, 1.5 seconds later: Games XOVERLAPPED completion handler```

Though like I said at the bottom of the OP, creating new threads for every export that uses XOVERLAPPED probably isn't the best idea, ideally we'd have some kind of task thread that we could queue these functions to (maybe we can use KernelState::dispatch_thread_ / dispatch_queue_ for this?)
